src/Lib.P2P/SecureCommunication/Psk1Stream.cs

```diff
@@ -0,0 +1,180 @@
+﻿using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Threading;
+using System.Threading.Tasks;
+using Lib.P2P.Cryptography;
+using Org.BouncyCastle.Crypto;
+using Org.BouncyCastle.Crypto.Engines;
+using Org.BouncyCastle.Crypto.Parameters;
+
+namespace Lib.P2P.SecureCommunication
+{
+    /// <summary>
+    ///   A duplex stream that is encrypted with a <see cref="PreSharedKey"/>.
+    /// </summary>
+    /// <remarks>
+    ///   The XSalsa20 cipher is used to encrypt the data.
+    /// </remarks>
+    /// <seealso href="https://github.com/libp2p/specs/blob/master/pnet/Private-Networks-PSK-V1.md"/>
+    public class Psk1Stream : Stream
+    {
+        private const int KeyBitLength = 256;
+        private const int NonceBitLength = 192;
+        private const int NonceByteLength = NonceBitLength / 8;
+
+        private Stream stream;
+        private PreSharedKey key;
+        private IStreamCipher readCipher;
+        private IStreamCipher writeCipher;
+
+        /// <summary>
+        ///   Creates a new instance of the <see cref="Psk1Stream"/> class. 
+        /// </summary>
+        /// <param name="stream">
+        ///   The source/destination of the unprotected stream.
+        /// </param>
+        /// <param name="key">
+        ///   The pre-shared 256-bit key for the private network of peers.
+        /// </param>
+        public Psk1Stream(Stream stream,
+            PreSharedKey key)
+        {
+            if (key.Length != KeyBitLength)
+                throw new Exception($"The pre-shared key must be {KeyBitLength} bits in length.");
+
+            this.stream = stream;
+            this.key = key;
+        }
+
+        private IStreamCipher WriteCipher
+        {
+            get
+            {
+                if (writeCipher == null)
+                {
+                    // Get a random nonce
+                    var nonce = new byte[NonceByteLength];
+                    using (var rng = RandomNumberGenerator.Create())
+                    {
+                        rng.GetBytes(nonce);
+                    }
+
+                    // Send the nonce to the remote
+                    stream.Write(nonce, 0, nonce.Length);
+
+                    // Create the cipher
+                    writeCipher = new XSalsa20Engine();
+                    writeCipher.Init(true, new ParametersWithIV(new KeyParameter(key.Value), nonce));
+                }
+
+                return writeCipher;
+            }
+        }
+
+        private IStreamCipher ReadCipher
+        {
+            get
+            {
+                if (readCipher == null)
+                {
+                    // Get the nonce from the remote.
+                    var nonce = new byte[NonceByteLength];
+                    for (int i = 0, n; i < NonceByteLength; i += n)
+                    {
+                        n = stream.Read(nonce, i, NonceByteLength - i);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Do not update the loop counter 'n' within the loop body.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)